cardinal/message/manager.go

```diff
@@ -79,13 +83,27 @@ func (m *Manager) GetMessageByID(id types.MessageID) types.Message {
 
 // GetMessageByName returns the message with the given name, if it exists.
 func (m *Manager) GetMessageByName(name string) (types.Message, bool) {
-	msg, ok := m.registeredMessages[name]
+	msg, ok := m.registeredMessagesByName[name]
+	return msg, ok
+}
+
+func (m *Manager) GetMessageByType(mType reflect.Type) (types.Message, bool) {
+	msg, ok := m.registeredMessagesByType[mType]
 	return msg, ok
 }
 
+func (m *Manager) RegisterMessageByType(mType reflect.Type, message types.Message) error {
```

Comment:
This is used by:

`RegisterMessage[In any, Out any](world *World, name string, opts ...message.MessageOption[In, Out]) error`

It's not used by GetMessage, that's another method, it's easy to get mixed up. The problem is similar to the one I described above. If I want to have a generic function to access a Registered value, that generic function cannot be a method by nature of being generic. 

If it can't be a method, then that means it can't access private variables so I have to sort of create a public route to the underlying method. 

As I stated previously I'm adapting the code with trade offs to allow for functions such as this: 
`RegisterMessage[In any, Out any](world *World, name string, opts ...message.MessageOption[In, Out]) error`
and this:
`GetMessage[In any, Out any](wCtx engine.Context) (*message.MessageType[In, Out], error)`

I'm definitely on board with changing the function signatures above or finding alternative ways to not have this access path. Very much open to your suggestions here.

Reply:
This is meant for `GetMessageByType` i believe. 

For the way to do this cleanly, see my code for RegisterComponent. 

If i understand correctly, the goal is to make this the one that is canonical once the refactor is finished? then this is the one who should be `RegisterMessage` and `RegisterMessageByName`/`RegisterMessages` above is the one that needs to be killed. Apologies, this whole "keeping both ways to keep the PR small" thing is confusing me on what i actually need to review, but you do you.

```
func RegisterComponent[T types.Component](w *World) error {
	if w.worldStage.Current() != worldstage.Init {
		return eris.Errorf(
			"engine state is %s, expected %s to register component",
			w.worldStage.Current(),
			worldstage.Init,
		)
	}

	compMetadata, err := component.NewComponentMetadata[T]()
	if err != nil {
		return err
	}

	err = w.componentManager.RegisterComponent(compMetadata)
	if err != nil {
		return err
	}

	return nil
}
```